core/graph/src/main/java/uk/gov/gchq/gaffer/graph/hook/FunctionAuthoriser.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright 2019 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package uk.gov.gchq.gaffer.graph.hook;
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonPropertyOrder;
+import com.fasterxml.jackson.databind.JsonNode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import uk.gov.gchq.gaffer.commonutil.exception.UnauthorisedException;
+import uk.gov.gchq.gaffer.exception.SerialisationException;
+import uk.gov.gchq.gaffer.jsonserialisation.JSONSerialiser;
+import uk.gov.gchq.gaffer.operation.OperationChain;
+import uk.gov.gchq.gaffer.store.Context;
+import uk.gov.gchq.koryphe.impl.predicate.And;
+import uk.gov.gchq.koryphe.impl.predicate.Not;
+import uk.gov.gchq.koryphe.impl.predicate.Or;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * The FunctionAuthoriser is a {@link GraphHook} which stops a user running
+ * Functions which have been banned. The Authoriser can be configured with a
+ * list of unauthorised function classes or patterns, or a list of authorised
+ * patterns to check against.
+ * <p>
+ * It should be noted that using the unauthorisedFunctions list will be more
+ * efficient than using any of the patterns.
+ */
+@JsonPropertyOrder(value = {"unauthorisedFunctions", "unauthorisedFunctionPatterns", "authorisedFunctionPatterns"})
+@JsonInclude(JsonInclude.Include.NON_DEFAULT)
+public class FunctionAuthoriser implements GraphHook {
+
+    private static final String ERROR_MESSAGE_PREFIX = "Operation chain contained an unauthorised function: ";
+    private static final Logger LOGGER = LoggerFactory.getLogger(FunctionAuthoriser.class);
+    private List<Class<? extends Function>> unauthorisedFunctions = new ArrayList<>();
+    private List<Pattern> authorisedFunctionPatterns = new ArrayList<>();
```

Comment:
Just wondering if it might be simpler to just call these white and black lists? As you refer to white and black lists in other parts of the code.

Reply:
I agree it's probably better to be consistent, although when creating this in JSON, it makes it more understandable in my opinion - especially given that there are two whitelists.